go/service/user.go

```diff
@@ -634,48 +633,17 @@ func (h *UserHandler) UserCard(ctx context.Context, arg keybase1.UserCardArg) (r
 
 func (h *UserHandler) SetUserBlocks(ctx context.Context, arg keybase1.SetUserBlocksArg) (err error) {
 	mctx := libkb.NewMetaContext(ctx, h.G())
-	defer mctx.TraceTimed(
-		fmt.Sprintf("UserHandler#SetUserBlocks(len=%d)", len(arg.Blocks)),
-		func() error { return err })()
-
-	type setBlockArg struct {
-		BlockUID string `json:"block_uid"`
-		Chat     *bool  `json:"chat,omitempty"`
-		Follow   *bool  `json:"follow,omitempty"`
-	}
-
-	for _, block := range arg.Blocks {
-		h.G().Log.CDebugf(ctx, "SetUserBlocks: adding block: %+v", block)
-	}
-	payloadBlocks := make([]setBlockArg, len(arg.Blocks))
-	uids := make([]keybase1.UID, len(arg.Blocks))
-	for i, v := range arg.Blocks {
-		uid := libkb.GetUIDByUsername(h.G(), v.Username)
-		payloadBlocks[i] = setBlockArg{
-			BlockUID: uid.String(),
-			Chat:     v.SetChatBlock,
-			Follow:   v.SetFollowBlock,
-		}
-		uids[i] = uid
-	}
-
-	payload := make(libkb.JSONPayload)
-	payload["blocks"] = payloadBlocks
-
-	apiArg := libkb.APIArg{
-		Endpoint:    "user/set_blocks",
-		JSONPayload: payload,
-		SessionType: libkb.APISessionTypeREQUIRED,
+	eng := engine.NewUserBlocksSet(h.G(), arg)
+	uis := libkb.UIs{
+		LogUI:     h.getLogUI(arg.SessionID),
+		SessionID: arg.SessionID,
 	}
-
-	_, err = mctx.G().API.Post(mctx, apiArg)
-
-	if err == nil {
-		h.cleanupAfterBlockChange(mctx, uids)
+	mctx = mctx.WithUIs(uis)
+	if err := engine.RunEngine2(mctx, eng); err != nil {
+		return err
 	}
-
-	return err
-
+	h.cleanupAfterBlockChange(mctx, eng.UIDs())
```

Comment:
ideally this would be in the engine as well, but it's also used in the other RPC, so it's fine for now.